From: SeongJae Park <sj@kernel.org>
Date: Mon, 17 Mar 2025 11:10:49 -0700
Subject: [PATCH] mm/damon: add DAMOS_FILTER_TYPE_ACTIVE

This could be useful at finding conflicting views between DAMON and LRU
placement logic.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 include/linux/damon.h    | 2 ++
 mm/damon/paddr.c         | 3 +++
 mm/damon/sysfs-schemes.c | 1 +
 3 files changed, 6 insertions(+)

diff --git a/include/linux/damon.h b/include/linux/damon.h
index 1ec1eb0b95ab..9c6beb2e8113 100644
--- a/include/linux/damon.h
+++ b/include/linux/damon.h
@@ -353,6 +353,7 @@ struct damos_stat {
  * @DAMOS_FILTER_TYPE_YOUNG:	Recently accessed pages.
  * @DAMOS_FILTER_TYPE_HUGEPAGE_SIZE:	Page is part of a hugepage.
  * @DAMOS_FILTER_TYPE_UNMAPPED:	Unmapped pages.
+ * @DAMOS_FILTER_TYPE_ACTIVE:	Page is in active LRU lists.
  * @DAMOS_FILTER_TYPE_ADDR:	Address range.
  * @DAMOS_FILTER_TYPE_TARGET:	Data Access Monitoring target.
  * @NR_DAMOS_FILTER_TYPES:	Number of filter types.
@@ -374,6 +375,7 @@ enum damos_filter_type {
 	DAMOS_FILTER_TYPE_YOUNG,
 	DAMOS_FILTER_TYPE_HUGEPAGE_SIZE,
 	DAMOS_FILTER_TYPE_UNMAPPED,
+	DAMOS_FILTER_TYPE_ACTIVE,
 	DAMOS_FILTER_TYPE_ADDR,
 	DAMOS_FILTER_TYPE_TARGET,
 	NR_DAMOS_FILTER_TYPES,
diff --git a/mm/damon/paddr.c b/mm/damon/paddr.c
index ebad6113b3f4..cced4e9b53fe 100644
--- a/mm/damon/paddr.c
+++ b/mm/damon/paddr.c
@@ -379,6 +379,9 @@ static bool damos_pa_filter_match(struct damos_filter *filter,
 	case DAMOS_FILTER_TYPE_UNMAPPED:
 		matched = !folio_mapped(folio) || !folio_raw_mapping(folio);
 		break;
+	case DAMOS_FILTER_TYPE_ACTIVE:
+		matched = folio_test_active(folio);
+		break;
 	default:
 		break;
 	}
diff --git a/mm/damon/sysfs-schemes.c b/mm/damon/sysfs-schemes.c
index 5e912c8fb3d1..3a072ce1335c 100644
--- a/mm/damon/sysfs-schemes.c
+++ b/mm/damon/sysfs-schemes.c
@@ -348,6 +348,7 @@ static const char * const damon_sysfs_scheme_filter_type_strs[] = {
 	"young",
 	"hugepage_size",
 	"unmapped",
+	"active",
 	"addr",
 	"target",
 };
-- 
2.39.5

